### PR TITLE
Improve error handling for underscores in odd places

### PR DESCRIPTION
**Pull request by DanilaFe**
Closes https://github.com/chapel-lang/chapel/issues/25002. Closes https://github.com/chapel-lang/chapel/issues/25001.

This PR takes exactly the steps I outlined in #25002. Specifically, it:

1. Allows `_` to make it from the parser to the AST, even in non-declaration contexts like `var x = _`.
2. Adds a post-parse check to detect `_` and warn about it, _except_ in places where it's allowed
3. Makes that post-parse check have a nicer error message :)

While there, it improves the error message for `for` in particular, where it explains what the user should do if they don't care about their identifier. Thus, one gets:

Screen Shot 2024-06-04 at 4 23 53 PM

That last line is specific to indexable loops.

Reviewed by @vasslitvinov -- thanks!

# Testing
- [x] paratest

### CODE REVIEW

**Review by vasslitvinov — APPROVED** [review]
Looks good!

File: test/errors/parsing/underscores.chpl [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import this.A as _;
+import this.A.{dummy as _};
+
+// Use statements can't rename limitations to '_'
+use A only dummy as _;
+
+// Underscores ARE allowed in tuple declarations...
+var (_, _) = (42, 42);
+// .. tuple assignments
+(x, y) = (42, 42);
```
Comment (cassella) [inline test/errors/parsing/underscores.chpl:30]: The test that underscores are allowed in tuple assignments doesn't include an underscore...

**Review by cassella — COMMENTED** [review]
This also resolved #24102, though that particular case isn't covered by the test.  Should I add it?

File: test/errors/parsing/underscores.chpl (hunk 1 above)
Comment (DanilaFe) [reply test/errors/parsing/underscores.chpl:30]: Thank you for pointing this out! I will need to adjust the test.

**Comment by DanilaFe** [comment]
> Should I add it?

Please feel free to open a pull request! Alternatively, if you'd rather I do it, I can add a test case myself.

File: test/errors/parsing/underscores.chpl (hunk 1 above)
Comment (DanilaFe) [reply test/errors/parsing/underscores.chpl:30]: https://github.com/chapel-lang/chapel/pull/25303 adjusts the test to actually use an underscore 😄

**Pull request merged**